flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/allocator/SlotSharingSlotAllocator.java

```diff
@@ -121,16 +133,13 @@ public Optional<VertexParallelismWithSlotSharing> determineParallelism(
                             slotSharingGroupParallelism.get(
                                     slotSharingGroup.getSlotSharingGroupId()));
 
-            final Iterable<ExecutionSlotSharingGroup> sharedSlotToVertexAssignment =
+            final List<ExecutionSlotSharingGroup> sharedSlotToVertexAssignment =
                     createExecutionSlotSharingGroups(vertexParallelism);
 
-            for (ExecutionSlotSharingGroup executionSlotSharingGroup :
-                    sharedSlotToVertexAssignment) {
-                final SlotInfo slotInfo = slotIterator.next();
-
-                assignments.add(
-                        new ExecutionSlotSharingGroupAndSlot(executionSlotSharingGroup, slotInfo));
-            }
+            SlotAssigner.AssignmentResult result =
+                    slotAssigner.assignSlots(freeSlots, sharedSlotToVertexAssignment);
+            assignments.addAll(result.assignments);
+            freeSlots = result.remainingSlots;
```

Comment:
Because `assignSlots` is called in a loop, and internally iterates over the `freeSlots`, the complexity here is suboptimal. Lower-priority slots are added over and over into the queue inside `assignSlots`, resulting in quadratic complexity.

I was thinking about making free slots part of `SlotAssigner` state (passed into constructor).

WDYT @dmvk ?

Reply:
🤔 Is there a reason to run this within the loop at all? Can we simply move this to after the loop?

Reply:
Fixed by restructuring the code and going through all the groups in on go: 2fc84d83884c1f342d46fec02d961d4f73eb4219
..4fa93e3756fad77cf2dbc92bf77614aa39cc28fe.